pkg/filter/wasm/wasm.go

```diff
@@ -0,0 +1,317 @@
+
+func init() {
+	for i := int32(1); i <= maxWasmCode; i++ {
+		results = append(results, wasmCodeToResult(i))
+	}
+	httppipeline.Register(&WasmFilter{})
+}
+
+type (
+	Spec struct {
+		MaxConcurrency int32  `yaml:"maxConcurrency" jsonschema:"required,minimum=1"`
+		Code           string `yaml:"code" jsonschema:"required"`
+		Timeout        string `yaml:"timeout" jsonschema:"required,format=duration"`
+		timeout        time.Duration
+	}
+
+	WasmFilter struct {
+		pipeSpec *httppipeline.FilterSpec
+		spec     *Spec
+
+		code   []byte
+		vmPool atomic.Value
+		chStop chan struct{}
+	}
+)
+
+// Kind returns the kind of WasmFilter.
+func (f *WasmFilter) Kind() string {
+	return Kind
+}
+
+// DefaultSpec returns the default spec of WasmFilter.
+func (f *WasmFilter) DefaultSpec() interface{} {
+	return &Spec{
+		MaxConcurrency: 10,
+		Timeout:        "100ms",
+	}
+}
+
+// Description returns the description of WasmFilter
+func (f *WasmFilter) Description() string {
+	return "WasmFilter implements a filter which runs web assembly"
+}
+
+// Results returns the results of WasmFilter.
+func (f *WasmFilter) Results() []string {
+	return results
+}
+
+func readWasmCodeFromFile(path string) ([]byte, error) {
+	f, e := os.Open(path)
+	if e != nil {
+		return nil, e
+	}
+	defer f.Close()
+	return io.ReadAll(f)
+}
+
+func readWasmCodeFromURL(url string) ([]byte, error) {
+	resp, e := http.DefaultClient.Get(url)
+	if e != nil {
+		return nil, e
+	}
+	defer resp.Body.Close()
+	return io.ReadAll(resp.Body)
+}
+
+func isURL(str string) bool {
+	// we only check the first a few bytes as str could be
+	// the LONG base64 encoded wasm code
+	for _, p := range []string{"http://", "https://"} {
+		if len(str) > len(p) {
+			s := strings.ToLower(str[:len(p)])
+			if strings.HasPrefix(s, p) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (f *WasmFilter) readWasmCode() ([]byte, error) {
+	if isURL(f.spec.Code) {
+		return readWasmCodeFromURL(f.spec.Code)
+	}
+	if _, e := os.Stat(f.spec.Code); e == nil {
+		return readWasmCodeFromFile(f.spec.Code)
+	}
+	return base64.StdEncoding.DecodeString(f.spec.Code)
+}
+
+func (f *WasmFilter) loadWasmCode() error {
+	code, e := f.readWasmCode()
+	if e != nil {
+		logger.Errorf("failed to load Wasm code: %v", e)
+		return e
+	}
+
+	if len(f.code) > 0 && bytes.Equal(f.code, code) {
+		return nil
+	}
+
+	p, e := NewWasmVMPool(f.spec.MaxConcurrency, code)
+	if e != nil {
+		logger.Errorf("failed to create Wasm VM pool: %v", e)
+		return e
+	}
+	f.code = code
+
+	f.vmPool.Store(p)
+	return nil
+}
+
+func (f *WasmFilter) watchWasmCode() {
+	var (
+		chWasm <-chan *string
+		syncer *cluster.Syncer
+		err    error
+	)
+
+	for {
+		c := f.pipeSpec.Super().Cluster()
+		syncer, err = c.Syncer(time.Minute)
+		if err == nil {
+			chWasm, err = syncer.Sync(c.Layout().WasmCodeEvent())
+			if err == nil {
+				break
+			}
+		}
+		logger.Errorf("failed to watch Wasm code event: %v", err)
+		select {
+		case <-time.After(10 * time.Second):
```

Comment:
Need to add `continue` here?

Reply:
no, it won't fall through to the next `case`.